File: QuickSelect.cs
using System;
using System.Collections.Generic;

namespace LeetCode.QuickSelect
{
    public static class QuickSelect
    {
        public static T Select<T>(IList<T> list, int left, int right, int k) where T : IComparable<T>
        {
            if (list == null)
            {
                throw new ArgumentNullException(nameof(list));
            }

            if (left < 0 || left > right || right >= list.Count || k > right - left + 1)
            {
                throw new ArgumentOutOfRangeException();
            }

            k += left - 1;
            var rand = new Random();
            while (left <= right)
            {
                var pivot = rand.Next(left, right + 1);
                pivot = Partition(list, left, right, pivot);
                if (pivot == k)
                {
                    return list[pivot];
                }
                else if (pivot > k)
                {
                    right = pivot - 1;
                }
                else
                {
                    left = pivot + 1;
                }
            }

            throw new ApplicationException();
        }

        private static int Partition<T>(IList<T> list, int left, int right, int pivot) where T : IComparable<T>
        {
            Swap(list, pivot, right);

            var store = left;
            for (var i = left; i < right; i++)
            {
                if (list[i].CompareTo(list[right]) < 0)
                {
                    Swap(list, store++, i);
                }
            }

            Swap(list, store, right);

            return store;
        }

        private static void Swap<T>(IList<T> list, int index1, int index2)
        {
            var temp = list[index1];
            list[index1] = list[index2];
            list[index2] = temp;
        }
    }
}

File: QuickSelect.java
public class QuickSelect {
    public static int select(int[] a, int k) {
        return select(a, k, 0, a.length - 1);
    }

    private static int select(int[] a, int k, int l, int r) {
        int p = medianThree(a, l, r);
        int i = l, j = r - 1;
        while (i < j) {
            while (a[++i] < p) ;
            while (a[--j] > p) ;
            if (i < j) exch(a, i, j);
        }
        exch(a, i, r - 1);
        if (k < i) return select(a, k, l, i - 1);
        else if (k == i) return a[k];
        else return select(a, k, i + 1, r);
    }

    private static  int medianThree(int[] a, int l, int r) {
        int m = l + (r - l) / 2;
        if (a[l] > a[m]) exch(a, l, m);
        if (a[l] > a[r]) exch(a, l, r);
        if (a[m] > a[r]) exch(a, m, r);
        exch(a, m, r - 1);
        return a[r - 1];
    }

    private static void exch(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
}
